compiler/rustc_parse/src/lexer/tokentrees.rs

```diff
@@ -17,10 +20,11 @@ pub(super) struct TokenTreesReader<'a> {
     /// The type and spans for all braces
     ///
     /// Used only for error recovery when arriving to EOF with mismatched braces.
-    matching_delim_spans: Vec<(Delimiter, Span, Span)>,
     last_unclosed_found_span: Option<Span>,
+
     /// Collect empty block spans that might have been auto-inserted by editors.
-    last_delim_empty_block_spans: FxHashMap<Delimiter, Span>,
+    empty_block_spans: FxHashMap<Span, Delimiter>,
+
     /// Collect the spans of braces (Open, Close). Used only
     /// for detecting if blocks are empty and only braces.
     matching_block_spans: Vec<(Span, Span)>,
```

Comment:
```suggestion
    diag: TokenTreesReaderDiag,
```
Could you move all these diagnostic-only fields to a separate structure?
(It could be passed as a whole to `report_suspicious_mismatch_block` as well.)